enhancements/api-review/custom-tags-aws.md

```diff
@@ -79,65 +101,373 @@ The userTags field is intended to be set at install time and is considered immut
+
+In the case where a user-defined tag is specified in the Infrastructure resource and
+1) There is already a tag with the same key present for AWS resource, the value from the AWS resource will be replaced.
+
+   For example,
+
+   Existing tag for AWS resource = `key_infra1 = value_comp1`
+
+   New tag request = `.spec.platformSpec.aws.resourceTags` has `key_infra1 = value_infra1`
+
+   Action = Existing tag for AWS resource is updated to reflect new value.
+
+   Final tag set to AWS resource = `key_infra1 = value_infra1`
+
+   Event action = An event is generated to notify user about the action status (success/failure) to update tags for the AWS resource.
+
+2) There is no tag with same key present, new tag is created for the AWS resource. In case of limit reached, a validation error is generated.
+
+   For example,
+
+   Existing tag for AWS resource = `key_infra1 = value_comp1`
+
+   New tag request = `.spec.platformSpec.aws.resourceTags` has `key_infra1 = value_infra1`
+
+   Action = A new tag is created for AWS resource.
+
+   Final tag set to AWS resource = `key_infra1 = value_infra1`
+
+   Event action = An event is generated to notify user about the action status (success/failure) to create tags for the AWS resource.
+
+#### Update tags scenarios
+Users can update the user-defined tags by editing `.spec.platformSpec.aws` of the `Infrastructure.config.openshift.io` type.
+On update of `resourceTags`, AWS resource is not created or restarted.
+
+In the case where a user-defined tag is specified in the Infrastructure resource and
+1) There is already a tag with the same key and value present for AWS resource, the tag for the AWS resource will not be updated.
+
+   For example,
+
+   Existing tag for AWS resource = `key_infra1 = value_update1`
+
+   New tag request = `.spec.platformSpec.aws.resourceTags` has `key_infra1 = value_update1`
+
+   Action = There is no update for AWS resource.
+
+   Final tag set to AWS resource = `key_infra1 = value_update1`
+
+   Event action = An event is generated to notify user about the action status (success/failure) to update tags for the AWS resource.
+
+2) There is already a tag with the same key and different value present for AWS resource, the user-define tag value for the AWS resource will be updated.
+
+   For example,
+
+   Existing tag for AWS resource = `key_infra1 = value_old`
+
+   New tag request = `.spec.platformSpec.aws.resourceTags` has `key_infra1 = value_update1`
+
+   Action = Existing tag for AWS resource is updated to reflect new value.
+
+   Final tag set to AWS resource = `key_infra1 = value_update1`
+
+   Event action = An event is generated to notify user about the action status (success/failure) to update tags for the AWS resource.
+
+#### Delete tags scenarios
+Tags are deleted when the user sets the user-defined tag value to an empty string in `.spec.platformSpec.aws.resourceTags`.
```

Comment:
Are we certain that there is no use case for a user wanting a tag with an empty value on a resource?

Reply:
Generally, cluster admins tend to set tags with default value set to empty string during creation of resources. Later based on the usecase, the values are updated in these tags or set to empty string. The OpenShift installer fails on validation for using empty strings.

@staebler , 
1) do you suggest to keep the behaviour same with updates? i.e, fail validation on empty strings?
2) is it better to consider value as an optional field and consider "nil"/null value field for deletion?
In case of cloud providers, the value of a tag can be set to empty string but cannot be null.

Reply:
My suggestion would be to separate the tags to delete entirely from the tags to maintain. That way it is clearer to the user and less likely from them to experience surprises.

Reply:
Do you think it is better to delete tags on AWS resources when removed from spec? This behavior will be consistent with usual CRUD-like object handling.

As tag update/delete can be overridden by entries in kube resource objects, making tags "unmanaged" wiithout deleting requires tricky methods to handle delete and maintain the limits for the number of tags. It is better to NOT support the following user story. 

```
- As a cluster administrator of OpenShift, I would like to be able to remove user-defined tags from `.spec.platformSpec.aws.resourceTags` without deleting for AWS resources.
  This enables, user-defined tags to be modified on AWS without being overridden by OpenShift operators. The user can adopt new 3rd party component to manage tags.
```
@JoelSpeed @staebler , Do you foresee any usecase that might require the above user story? Third-party tools operate without any OpenShift specific awareness of tag management, the user story does not seem to add much benefit. In this case, user would be required to delete the tag from Openshift and re-tag AWS resources using the third-party tools.

Reply:
I agree with Joel. I don't think that we can support removal of AWS tags when a tag is removed from the Infrastructure's `resourceTags`.

Reply:
Yes, from UX perspective, it definitely helps user if there is more detailed and discrete tag information.

Following reasons user may find different lists confusing
1) There is no deterministic time to remove the entry from delete list.
2) Conflicting entries are possible which adds confusion to user if the delete list tag entry is not removed.
3) Conflict scenarios with override is possible. e.g: One of the scenarios is, when the kube resource overriding entry is removed and there is no entry in maintain list and delete list.

The following might help user to track tag history or requests,
1) For tag history, velero backups can be used to save.
2) For monitoring operations on tags, events are generated for tag add/update/delete requests.
Also, based on specific events, alerts can be triggered.
3) For monitoring tag changes on AWS resource, AWS CloudWatch Events can be explored.

It is less complex to remove entry from spec for deletion and let user opt for one of the above methods to track tags.

@JoelSpeed, Do you think it is ok to go ahead with the above methods for tracking tag?

Reply:
Based on the discussion with @JoelSpeed and @staebler ,  the change is going to be ambiguous to the user. It may also involve versioning the tags. It is decided to not consider in this enhancement.